# INFO-F-422 -  Statistical Foundations of Machine Learning 

### Student 1 - __[Ghita.Ait.Ouhmane@ulb.ac.be](mailto:Ghita.Ait.Ouhmane@ulb.ac.be) - Student ID 453422__
### Student 2 - __[Ha.Duong@ulb.ac.be](mailto:Ha.Duong@ulb.ac.be) - Student ID 463350__
### Student 3 - __[Maxime.Langlet@ulb.ac.be](mailto:Maxime.Langlet@ulb.ac.be) - Student ID 427821__

### Video presentation: www.youtube.com/abcd1234

## Statistical foundations of Machine Learning - Project 2020-2021


# Introduction


In [1]:
# install the packages and libraries
install.packages("e1071")
install.packages("tidyverse")
install.packages("pillar")
library(rpart)
library(e1071)
library(rpart.plot)##??
library(randomForest)

Installing package into 'C:/Users/duong/OneDrive/Documents/R/win-library/4.0'
(as 'lib' is unspecified)



package 'e1071' successfully unpacked and MD5 sums checked


Warning message:
"cannot remove prior installation of package 'e1071'"
Warning message in file.copy(savedcopy, lib, recursive = TRUE):
"problème lors de la copie de C:\Users\duong\OneDrive\Documents\R\win-library\4.0\00LOCK\e1071\libs\x64\e1071.dll vers C:\Users\duong\OneDrive\Documents\R\win-library\4.0\e1071\libs\x64\e1071.dll: Permission denied"
Warning message:
"restored 'e1071'"



The downloaded binary packages are in
	C:\Users\duong\AppData\Local\Temp\RtmpYZ0r8T\downloaded_packages


Installing package into 'C:/Users/duong/OneDrive/Documents/R/win-library/4.0'
(as 'lib' is unspecified)



package 'tidyverse' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\duong\AppData\Local\Temp\RtmpYZ0r8T\downloaded_packages


Installing package into 'C:/Users/duong/OneDrive/Documents/R/win-library/4.0'
(as 'lib' is unspecified)



package 'pillar' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\duong\AppData\Local\Temp\RtmpYZ0r8T\downloaded_packages


ERROR: Error in library(rpart.plot): there is no package called 'rpart.plot'


'''date_recorded
num_private
funder
installer
wpt_name
subvillage
region? (code 27, region 21, why?)
lga
ward
recorded_by
scheme_name
payment
quantity
waterpoint_type_group'''

In [62]:
# read the datasets

sub_format <- read.csv("database/Pump_it_Up_Data_Mining_the_Water_Table_-_Submission_format.csv", header=TRUE)
set_val_test <- read.csv("database/Pump_it_Up_Data_Mining_the_Water_Table_-_Test_set_values.csv", header=TRUE)
set_labels <- read.csv("database/Pump_it_Up_Data_Mining_the_Water_Table_-_Training_set_labels.csv", header=TRUE)
dataset <- read.csv("database/Pump_it_Up_Data_Mining_the_Water_Table_-_Training_set_values.csv", header=TRUE)

# print the number of values of the Y
choices <- length(unique(set_labels$status_group))
choices
# set_labels
# dim(set_labels)

[1] 3

# Data preprocessing

## Missing value imputation




In [16]:
#On some distribution, when importing the dataset, some NA values are replaced by empty strings
#so we set them all tho NA
data1<-replace(dataset,dataset=='',NA)# Replaced to NA

#We then need to see whether there are NA values present in the dataset
#We can do that by removing them using na.omit and compare the dimension of the result with the input
dim(data1)
dim(na.omit(data1))


[1] 59400    40

[1] 27813    40

We can see that the number of rows decreased drastically, so we have to replace those NA values

## Feature selection
We now have to see which features will be kept, and which one introduce redundancy, or whose information aren't relevant to the problem.<br>

///this is just a draft<br>
Looking at the column types, we come accross two types of values : numerical and categorical. For the categorical values, a new encoding is required (like the one hot encoding for example) to obtain numerical values. However, the more unique values for a given columns, the more columns will be added to the dataset, so the features to keep have to take this into account.

In [33]:
head(data1)

id    amount_tsh date_recorded funder              gps_height installer   
1 69572 6000       2011-03-14    Roman               1390       Roman       
2  8776    0       2013-03-06    Grumeti             1399       GRUMETI     
3 34310   25       2013-02-25    Lottery Club         686       World vision
4 67743    0       2013-01-28    Unicef               263       UNICEF      
5 19728    0       2011-07-13    Action In A            0       Artisan     
6  9944   20       2011-03-13    Mkinga Distric Coun    0       DWE         
  longitude latitude   wpt_name             num_private ... payment_type
1 34.93809   -9.856322 none                 0           ... annually    
2 34.69877   -2.147466 Zahanati             0           ... never pay   
3 37.46066   -3.821329 Kwa Mahundi          0           ... per bucket  
4 38.48616  -11.155298 Zahanati Ya Nanyumbu 0           ... never pay   
5 31.13085   -1.825359 Shuleni              0           ... never pay   
6 39.17280   -4.765587 Tajiri               0           ... per bucket  
  water_quality quality_group quantity     quantity_group source              
1 soft          good          enough       enough         spring              
2 soft          good          insufficient insufficient   rainwater harvesting
3 soft          good          enough       enough         dam                 
4 soft          good          dry          dry            machine dbh         
5 soft          good          seasonal     seasonal       rainwater harvesting
6 salty         salty         enough       enough         other               
  source_type          source_class waterpoint_type            
1 spring               groundwater  communal standpipe         
2 rainwater harvesting surface      communal standpipe         
3 dam                  surface      communal standpipe multiple
4 borehole             groundwater  communal standpipe multiple
5 rainwater harvesting surface      communal standpipe         
6 other                unknown      communal standpipe multiple
  waterpoint_type_group
1 communal standpipe   
2 communal standpipe   
3 communal standpipe   
4 communal standpipe   
5 communal standpipe   
6 communal standpipe

- The first feature we can look at is the date recorded. According to the description of the features, the row date_recorded corresponds to the date the row was entered. So it is not relevant at all to our probem. We can then remove it from the dataset.
- The next fetaure is num_private. There is no description of it vailable on the datadriven website, so we can try looking at the values it holds : 

In [26]:
#All the values of this column present in this dataset
unique(data1$num_private) 
#Frequency of each value
(table(data1$num_private))

[1]    0   39    5   45    6    3  698   32   15    7   25  102    1   93   14
[16]   34  120   17  213   47    8   41   80  141   20   35  131    4   22   11
[31]   87   61   65  136    2  180   38   62    9   16   23   42   24   12  668
[46]  672   58  150  280  160   50 1776   30   27   10   94   26  450  240  755
[61]   60  111  300   55 1402


    0     1     2     3     4     5     6     7     8     9    10    11    12 
58643    73    23    27    20    46    81    26    46     4     4    11     5 
   14    15    16    17    20    22    23    24    25    26    27    30    32 
    3    35     8    17    14     6     1     5    12     2     2     1    40 
   34    35    38    39    41    42    45    47    50    55    58    60    61 
   10     1     4    30    10     1    36    21     5     1     4     1     3 
   62    65    80    87    93    94   102   111   120   131   136   141   150 
    1    22    15     1    28     1    20     1     7     1     1     1     6 
  160   180   213   240   280   300   450   668   672   698   755  1402  1776 
    1     1     1     1     1     1     1     1     1     1     1     1     1 

We see that more than half the values are zero, which could suggest the value is just missing. Anyway, the values are not diversified enough so we remove that column from our dataset.

In [37]:
#All the values of this column present in this dataset
length(unique(data1$funder)) 
length(unique(data1$installer))
length(unique(data1$wpt_name)) 
length(unique(data1$subvillage)) 

[1] 1898

[1] 2146

- The funder feature is also categorical. Looking at the number of unique values, there are 1898. Encoding this using one hot is not a good idea since it will introduce too many new parameters and will probably have a huge impact on the speed and accuracy of our future models.
- The same observation can be made for the installer, wpt_name and subvillage features.
- We can look at the following features together : region, region_code, lga, district code and ward :

In [51]:
length(unique(data1$lga)) 
length(unique(data1$ward))

#All the values of this column present in this dataset
length(unique(data1$region)) 
#Frequency of each value
(table(data1$region))

#All the values of this column present in this dataset
length(unique(data1$region_code)) 
#Frequency of each value
(table(data1$region_code))

#All the values of this column present in this dataset
length(unique(data1$district_code)) 
#Frequency of each value
(table(data1$district_code))

[1] 125

[1] 2092

[1] 21


       Arusha Dar es Salaam        Dodoma        Iringa        Kagera 
         3350           805          2201          5294          3316 
       Kigoma   Kilimanjaro         Lindi       Manyara          Mara 
         2816          4379          1546          1583          1969 
        Mbeya      Morogoro        Mtwara        Mwanza         Pwani 
         4639          4006          1730          3102          2635 
        Rukwa        Ruvuma     Shinyanga       Singida        Tabora 
         1808          2640          4982          2093          1959 
        Tanga 
         2547 

[1] 27


   1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
2201 3024 4379 2513 4040 1609  805  300  390 2640 5300 4639 2093 1979 1808 2816 
  17   18   19   20   21   24   40   60   80   90   99 
5011 3324 3047 1969 1583  326    1 1025 1238  917  423 

[1] 20


    0     1     2     3     4     5     6     7     8    13    23    30    33 
   23 12203 11173  9998  8999  4356  4074  3343  1043   391   293   995   874 
   43    53    60    62    63    67    80 
  505   745    63   109   195     6    12 

- We can right away remove features lga and ward, since they are both categorical and the number of unqiue valeus is very big.
## what conclusion for the 3 others ? 

In [57]:
length(unique(data1$scheme_name)) 

[1] 2697

- For the scheme_name, we see that we have 2967 different categorical values, wo we will drop this feature too.
## why did me remove payment when only 7 values ? 

In [59]:
length(unique(data1$payment)) 
(table(data1$payment))

[1] 7


            never pay                 other          pay annually 
                25348                  1054                  3642 
          pay monthly        pay per bucket pay when scheme fails 
                 8300                  8985                  3914 
              unknown 
                 8157 

In [64]:
length(unique(data1$quantity)) 
(table(data1$quantity))
length(unique(data1$quantity_group)) 
(table(data1$quantity_group))

length(unique(data1$waterpoint_type)) 
(table(data1$waterpoint_type))
length(unique(data1$waterpoint_type_group)) 
(table(data1$waterpoint_type_group))

length(unique(data1$recorded_by)) 
(table(data1$recorded_by))

[1] 5


         dry       enough insufficient     seasonal      unknown 
        6246        33186        15129         4050          789 

[1] 5


         dry       enough insufficient     seasonal      unknown 
        6246        33186        15129         4050          789 

[1] 7


              cattle trough          communal standpipe 
                        116                       28522 
communal standpipe multiple                         dam 
                       6103                           7 
                  hand pump             improved spring 
                      17488                         784 
                      other 
                       6380 

[1] 6


     cattle trough communal standpipe                dam          hand pump 
               116              34625                  7              17488 
   improved spring              other 
               784               6380 

[1] 1


GeoData Consultants Ltd 
                  59400 

- The two features quantity and quantity_group are identical, so we only need to keep one of them. We removed quantity
- For the fetaures waterpoint_type and waterpoint_type_group, we see that the values are similar expect for communal standpipe that is divided into "communal standpipe" and "communal standpipe multiplle" for the waterpoint_type feature. So we chose to keep this feature since it gives a little more variability to the data.
- Lookign at the recorded_by feature, we see that it only has one value which is "GeoData Consultants Ltd", so it doesn't bring any information to the dataset. It has been then removed.

In [65]:
# Delete the columns in cleaning based on the variance, the relevance of the information, etc.
cleaning <- c("date_recorded", "num_private", "funder", "installer", "wpt_name", "subvillage", "lga", "ward", 
              "region_code", "scheme_name", "payment", "quantity", "waterpoint_type_group", "recorded_by")
X<-dataset[,setdiff(colnames(dataset),cleaning)]
X<-replace(X,X=='',NA)#Replaced to NA
X[20:26]

id    amount_tsh gps_height longitude latitude   basin                  
1  69572 6000       1390       34.93809   -9.856322 Lake Nyasa             
2   8776    0       1399       34.69877   -2.147466 Lake Victoria          
3  34310   25        686       37.46066   -3.821329 Pangani                
4  67743    0        263       38.48616  -11.155298 Ruvuma / Southern Coast
5  19728    0          0       31.13085   -1.825359 Lake Victoria          
6   9944   20          0       39.17280   -4.765587 Pangani                
7  19816    0          0       33.36241   -3.766365 Internal               
8  54551    0          0       32.62062   -4.226198 Lake Tanganyika        
9  53934    0          0       32.71110   -5.146712 Lake Tanganyika        
10 46144    0          0       30.62699   -1.257051 Lake Victoria          
   region    district_code population public_meeting ... management
1  Iringa     5            109        True           ... vwc       
2  Mara       2            280        NA             ... wug       
3  Manyara    4            250        True           ... vwc       
4  Mtwara    63             58        True           ... vwc       
5  Kagera     1              0        True           ... other     
6  Tanga      8              1        True           ... vwc       
7  Shinyanga  3              0        True           ... vwc       
8  Shinyanga  3              0        True           ... wug       
9  Tabora     6              0        True           ... vwc       
10 Kagera     1              0        True           ... vwc       
   management_group payment_type water_quality quality_group quantity_group
1  user-group       annually     soft          good          enough        
2  user-group       never pay    soft          good          insufficient  
3  user-group       per bucket   soft          good          enough        
4  user-group       never pay    soft          good          dry           
5  other            never pay    soft          good          seasonal      
6  user-group       per bucket   salty         salty         enough        
7  user-group       never pay    soft          good          enough        
8  user-group       unknown      milky         milky         enough        
9  user-group       never pay    salty         salty         seasonal      
10 user-group       never pay    soft          good          enough        
   source               source_type          source_class
1  spring               spring               groundwater 
2  rainwater harvesting rainwater harvesting surface     
3  dam                  dam                  surface     
4  machine dbh          borehole             groundwater 
5  rainwater harvesting rainwater harvesting surface     
6  other                other                unknown     
7  machine dbh          borehole             groundwater 
8  shallow well         shallow well         groundwater 
9  machine dbh          borehole             groundwater 
10 shallow well         shallow well         groundwater 
   waterpoint_type            
1  communal standpipe         
2  communal standpipe         
3  communal standpipe multiple
4  communal standpipe multiple
5  communal standpipe         
6  communal standpipe multiple
7  hand pump                  
8  hand pump                  
9  hand pump                  
10 hand pump

We initially had 40 features. After removing the ones mentionned above, we are left with 26. 

## need to justify those we didnt remove ? 


## Feature engineering



Now that we have already removed some features, we can put the remainig ones in the correct format.<br>
For numerical values, we will replace all the NAs with the median value on that column.<br>
For categorical data, replace them with the most frequent value.

In [67]:
replace_na_with_median_value<-function(vec) {
    median_vec<-median(vec,na.rm=T)
    vec[is.na(vec)]<-median_vec
    vec
}

In [68]:
#distric code 0 ?? a demander
#population 0 -> mean ???
#public_meeting -> plus frequent
#scheme_management -> frequent
#permit -> plus frequent
#construction_year -> int(mean)
#dim(data.frame(X$permit))

#Replacing NA values in categorical features
table(X$permit)
X$permit[is.na(X$permit)] <- tail(names(sort(table(X$permit))), 1)
table(X$permit)

table(X$scheme_management)
X$scheme_management[is.na(X$scheme_management)] <- tail(names(sort(table(X$scheme_management))), 1)
table(X$scheme_management)

table(X$public_meeting)
X$public_meeting[is.na(X$public_meeting)] <- tail(names(sort(table(X$public_meeting))), 1)
table(X$public_meeting)


False  True 
17492 41908 


False  True 
17492 41908 


         Company             None            Other       Parastatal 
            1061                1              766             1680 
Private operator              SWC            Trust              VWC 
            1063               97               72            40670 
 Water authority      Water Board              WUA              WUG 
            3153             2748             2883             5206 


         Company             None            Other       Parastatal 
            1061                1              766             1680 
Private operator              SWC            Trust              VWC 
            1063               97               72            40670 
 Water authority      Water Board              WUA              WUG 
            3153             2748             2883             5206 


False  True 
 5055 54345 


False  True 
 5055 54345 

In [75]:
#For construction_year, some rows have a value 0, before replacing it , we first set 0 to NA so we can use is.na
X$construction_year[X$construction_year==0] <- NA
X$construction_year[is.na(X$construction_year)] <- median(X$construction_year,na.rm=T)
X$construction_year <- as.integer(X$construction_year)

In [77]:
#Same for longitude
X$longitude[X$longitude==0] <- NA
X$longitude[is.na(X$longitude)] <- mean(X$longitude,na.rm=T)

#X$latitude[X$latitude==0] <- NA
#X$latitude[is.na(X$latitude)] <- mean(X$latitude,na.rm=T)
#sum(X$latitude==-0.00000002)

In [78]:
#For permit, the categorical values are bolleans true-false so we can encode them on only one column with 0 and 1, same goes
#for public meeting
X$permit[X$permit=='True'] <- 1
X$permit[X$permit=='False'] <- 0
X$permit <- as.integer(X$permit)

X$public_meeting[X$public_meeting=='True'] <- 1
X$public_meeting[X$public_meeting=='False'] <- 0
X$public_meeting <- as.integer(X$public_meeting)

In [79]:
summary(X)
X[1:20]

       id          amount_tsh         gps_height       longitude    
 Min.   :    0   Min.   :     0.0   Min.   : -90.0   Min.   :29.61  
 1st Qu.:18520   1st Qu.:     0.0   1st Qu.:   0.0   1st Qu.:33.35  
 Median :37062   Median :     0.0   Median : 369.0   Median :35.15  
 Mean   :37115   Mean   :   317.7   Mean   : 668.3   Mean   :35.15  
 3rd Qu.:55657   3rd Qu.:    20.0   3rd Qu.:1319.2   3rd Qu.:37.18  
 Max.   :74247   Max.   :350000.0   Max.   :2770.0   Max.   :40.35  
    latitude          basin              region          district_code  
 Min.   :-11.649   Length:59400       Length:59400       Min.   : 0.00  
 1st Qu.: -8.541   Class :character   Class :character   1st Qu.: 2.00  
 Median : -5.022   Mode  :character   Mode  :character   Median : 3.00  
 Mean   : -5.706                                         Mean   : 5.63  
 3rd Qu.: -3.326                                         3rd Qu.: 5.00  
 Max.   :  0.000                                         Max.   :80.00  
   pop

id    amount_tsh gps_height longitude latitude    
1     69572 6000       1390       34.93809   -9.85632177
2      8776    0       1399       34.69877   -2.14746569
3     34310   25        686       37.46066   -3.82132853
4     67743    0        263       38.48616  -11.15529772
5     19728    0          0       31.13085   -1.82535885
6      9944   20          0       39.17280   -4.76558728
7     19816    0          0       33.36241   -3.76636472
8     54551    0          0       32.62062   -4.22619802
9     53934    0          0       32.71110   -5.14671181
10    46144    0          0       30.62699   -1.25705061
11    49056    0         62       39.20952   -7.03413939
12    50409  200       1062       35.77026  -10.57417468
13    36957    0          0       33.79811   -3.29019380
14    50495    0       1368       37.09257   -3.18178330
15    53752    0          0       34.36407   -3.62933335
16    61848    0       1645       31.44412   -8.27496163
17    48451  500       1703       34.64244   -9.10618458
18    58155    0       1656       34.56927   -9.08551497
19    34169    0       1162       32.92015   -1.94786801
20    18274  500       1763       34.50897   -9.89441250
21    48375  200       2216       34.47343   -9.59498965
22     6091    0          0       35.14967   -0.00000002
23    58500    0       1510       34.58690   -8.98001429
24    37862  500        672       37.94003   -4.12559468
25    51058    0       1645       29.74707   -4.49556758
26    22308    0       1273       37.42275   -3.31753648
27    55012  500        200       39.37078   -9.94253166
28    20145    0          0       31.10426   -1.73536124
29    19685    0       1443       37.61113   -3.26352630
30    69124    0       1256       37.06169   -3.19198926
...   ...   ...        ...        ...       ...         
59371 14796   200      1154       30.05873   -4.90263299
59372 20387     0         0       33.58641   -3.04036855
59373 29940     0         0       32.05855   -5.05210104
59374 15233  5000      1137       32.12502   -6.78435778
59375 49651     0         0       35.14967   -0.00000002
59376 50998 40000        54       39.04694   -5.06149401
59377 34716     0      1581       34.82104   -5.07625759
59378 43986     0         0       33.43701   -4.42359381
59379 38067   500      2101       34.62039   -9.35840452
59380 58255     0         0       31.61328   -1.82022635
59381 30647  6000      1439       34.94210   -9.84919761
59382 67885     0         0       38.83500   -4.88020370
59383 47002     6      1383       37.45476   -3.32359911
59384 44616     0        11       39.62601  -10.05749942
59385 72148     0         0       30.66780   -2.48370953
59386 34473   500      1327       33.95168   -2.02185406
59387 34952     0      1786       31.73879   -8.53201255
59388 26640   100        25       39.17648   -6.95709837
59389 72559     0      1414       30.69240   -3.59382702
59390 30410     0       783       30.64649   -7.36541790
59391 13677     0      1715       31.37085   -8.25816008
59392 44885     0       540       38.04407   -4.27221758
59393 40607     0         0       33.00944   -8.52088818
59394 48348     0         0       33.86685   -4.28740983
59395 11164   500       351       37.63405   -6.12482968
59396 60739    10      1210       37.16981   -3.25384746
59397 27263  4700      1212       35.24999   -9.07062880
59398 37057     0         0       34.01709   -8.75043433
59399 31282     0         0       35.86132   -6.37857327
59400 26348     0       191       38.10405   -6.74746425
      basin                   region        district_code population
1     Lake Nyasa              Iringa         5             109      
2     Lake Victoria           Mara           2             280      
3     Pangani                 Manyara        4             250      
4     Ruvuma / Southern Coast Mtwara        63              58      
5     Lake Victoria           Kagera         1               0      
6     Pangani                 Tanga          8            

In [15]:
names(X)

[1] "id"                    "amount_tsh"            "gps_height"           
 [4] "longitude"             "latitude"              "basin"                
 [7] "region"                "district_code"         "population"           
[10] "public_meeting"        "scheme_management"     "permit"               
[13] "construction_year"     "extraction_type"       "extraction_type_group"
[16] "extraction_type_class" "management"            "management_group"     
[19] "payment_type"          "water_quality"         "quality_group"        
[22] "quantity_group"        "source"                "source_type"          
[25] "source_class"          "waterpoint_type"

In [80]:
#one hot conversion
#install.packages('dummies')
library(dummies)

#Now that all fetaures have the correct format, we can encode categorical variables into classes with values 0-1 using
#the one hot encoding
variables_to_keep<-names(X)
X_onehot <- dummy.data.frame(X[,variables_to_keep], sep="_")

dummies-1.5.6 provided by Decision Patterns


Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"
Warning message in model.matrix.default(~x - 1, model.frame(~x - 

In [17]:
#names(data_factor_onehot)
X_onehot

id    amount_tsh gps_height longitude latitude     basin_Internal
1     69572 6000       1390       34.93809   -9.85632177 0             
2      8776    0       1399       34.69877   -2.14746569 0             
3     34310   25        686       37.46066   -3.82132853 0             
4     67743    0        263       38.48616  -11.15529772 0             
5     19728    0          0       31.13085   -1.82535885 0             
6      9944   20          0       39.17280   -4.76558728 0             
7     19816    0          0       33.36241   -3.76636472 1             
8     54551    0          0       32.62062   -4.22619802 0             
9     53934    0          0       32.71110   -5.14671181 0             
10    46144    0          0       30.62699   -1.25705061 0             
11    49056    0         62       39.20952   -7.03413939 0             
12    50409  200       1062       35.77026  -10.57417468 0             
13    36957    0          0       33.79811   -3.29019380 1             
14    50495    0       1368       37.09257   -3.18178330 0             
15    53752    0          0       34.36407   -3.62933335 1             
16    61848    0       1645       31.44412   -8.27496163 0             
17    48451  500       1703       34.64244   -9.10618458 0             
18    58155    0       1656       34.56927   -9.08551497 0             
19    34169    0       1162       32.92015   -1.94786801 0             
20    18274  500       1763       34.50897   -9.89441250 0             
21    48375  200       2216       34.47343   -9.59498965 0             
22     6091    0          0       35.14967   -0.00000002 0             
23    58500    0       1510       34.58690   -8.98001429 0             
24    37862  500        672       37.94003   -4.12559468 0             
25    51058    0       1645       29.74707   -4.49556758 0             
26    22308    0       1273       37.42275   -3.31753648 0             
27    55012  500        200       39.37078   -9.94253166 0             
28    20145    0          0       31.10426   -1.73536124 0             
29    19685    0       1443       37.61113   -3.26352630 0             
30    69124    0       1256       37.06169   -3.19198926 0             
...   ...   ...        ...        ...       ...          ...           
59371 14796   200      1154       30.05873   -4.90263299 0             
59372 20387     0         0       33.58641   -3.04036855 0             
59373 29940     0         0       32.05855   -5.05210104 0             
59374 15233  5000      1137       32.12502   -6.78435778 0             
59375 49651     0         0       35.14967   -0.00000002 0             
59376 50998 40000        54       39.04694   -5.06149401 0             
59377 34716     0      1581       34.82104   -5.07625759 1             
59378 43986     0         0       33.43701   -4.42359381 1             
59379 38067   500      2101       34.62039   -9.35840452 0             
59380 58255     0         0       31.61328   -1.82022635 0             
59381 30647  6000      1439       34.94210   -9.84919761 0             
59382 67885     0         0       38.83500   -4.88020370 0             
59383 47002     6      1383       37.45476   -3.32359911 0             
59384 44616     0        11       39.62601  -10.05749942 0             
59385 72148     0         0       30.66780   -2.48370953 0             
59386 34473   500      1327       33.95168   -2.02185406 0             
59387 34952     0      1786       31.73879   -8.53201255 0             
59388 26640   100        25       39.17648   -6.95709837 0             
59389 72559     0      1414       30.69240   -3.59382702 0             
59390 30410     0       783       30.64649   -7.36541790 0             
59391 13677     0      1715       31.37085   -8.25816008 0             
59392 44885     0       540       38.04407   -4.27221758 0             
59393 40607     0         0       33.00944   -8.52088818 0             
59394 48348     0         0       33.86685   -

In [84]:
vars <- c("id", "status_group") ##???
labels_onehot <- dummy.data.frame(set_labels[,vars], sep="_")

#Renaming the encoded features in the labels dataframe
names(labels_onehot)[names(labels_onehot) == "status_group_functional"] <- "functional"
names(labels_onehot)[names(labels_onehot) == "status_group_functional needs repair"] <- "functional_needs_repair"
names(labels_onehot)[names(labels_onehot) == "status_group_non functional"] <- "non_functional"
labels_onehot

#set_labels$status_group[set_labels$status_group=='functional']<-0
#set_labels$status_group[set_labels$status_group=='functional needs repair']<-0.5
#set_labels$status_group[set_labels$status_group=='non functional']<-1

#set_labels$status_group <- as.double(set_labels$status_group)
#set_labels

Warning message in model.matrix.default(~x - 1, model.frame(~x - 1), contrasts = FALSE):
"non-list contrasts argument ignored"


id    functional functional_needs_repair non_functional
1     69572 1          0                       0             
2      8776 1          0                       0             
3     34310 1          0                       0             
4     67743 0          0                       1             
5     19728 1          0                       0             
6      9944 1          0                       0             
7     19816 0          0                       1             
8     54551 0          0                       1             
9     53934 0          0                       1             
10    46144 1          0                       0             
11    49056 1          0                       0             
12    50409 1          0                       0             
13    36957 1          0                       0             
14    50495 1          0                       0             
15    53752 1          0                       0             
16    61848 1          0                       0             
17    48451 0          0                       1             
18    58155 0          0                       1             
19    34169 0          1                       0             
20    18274 1          0                       0             
21    48375 1          0                       0             
22     6091 1          0                       0             
23    58500 0          1                       0             
24    37862 1          0                       0             
25    51058 1          0                       0             
26    22308 0          1                       0             
27    55012 1          0                       0             
28    20145 1          0                       0             
29    19685 0          0                       1             
30    69124 1          0                       0             
...   ...   ...        ...                     ...           
59371 14796 1          0                       0             
59372 20387 0          0                       1             
59373 29940 0          0                       1             
59374 15233 1          0                       0             
59375 49651 1          0                       0             
59376 50998 1          0                       0             
59377 34716 0          0                       1             
59378 43986 0          0                       1             
59379 38067 1          0                       0             
59380 58255 1          0                       0             
59381 30647 0          0                       1             
59382 67885 0          0                       1             
59383 47002 0          0                       1             
59384 44616 1          0                       0             
59385 72148 0          0                       1             
59386 34473 1          0                       0             
59387 34952 1          0                       0             
59388 26640 1          0                       0             
59389 72559 1          0                       0             
59390 30410 1          0                       0             
59391 13677 1          0                       0             
59392 44885 0          0                       1             
59393 40607 0          0                       1             
59394 48348 1          0                       0             
59395 11164 0          0                       1             
59396 60739 1          0                       0             
59397 27263 1          0                       0             
59398 37057 1          0                       0             
59399 31282 1          0                       0             
59400 26348 1          0                       0

In [86]:
Y<-labels_onehot[,setdiff(colnames(labels_onehot),"id")]
Y<-set_labels[2]
Y[Y == "functional needs repair"] <- "functional_needs_repair"
Y[Y == "non functional"] <- "non_functional"
Y

status_group           
1     functional             
2     functional             
3     functional             
4     non_functional         
5     functional             
6     functional             
7     non_functional         
8     non_functional         
9     non_functional         
10    functional             
11    functional             
12    functional             
13    functional             
14    functional             
15    functional             
16    functional             
17    non_functional         
18    non_functional         
19    functional_needs_repair
20    functional             
21    functional             
22    functional             
23    functional_needs_repair
24    functional             
25    functional             
26    functional_needs_repair
27    functional             
28    functional             
29    non_functional         
30    functional             
...   ...                    
59371 functional             
59372 non_functional         
59373 non_functional         
59374 functional             
59375 functional             
59376 functional             
59377 non_functional         
59378 non_functional         
59379 functional             
59380 functional             
59381 non_functional         
59382 non_functional         
59383 non_functional         
59384 functional             
59385 non_functional         
59386 functional             
59387 functional             
59388 functional             
59389 functional             
59390 functional             
59391 functional             
59392 non_functional         
59393 non_functional         
59394 functional             
59395 non_functional         
59396 functional             
59397 functional             
59398 functional             
59399 functional             
59400 functional

In [87]:
#Removing id from the dataset, since the training dtaa and labels are both ordered in the same way
X_final<-X_onehot[,setdiff(colnames(X_onehot),"id")]
X_final

amount_tsh gps_height longitude latitude     basin_Internal
1     6000       1390       34.93809   -9.85632177 0             
2        0       1399       34.69877   -2.14746569 0             
3       25        686       37.46066   -3.82132853 0             
4        0        263       38.48616  -11.15529772 0             
5        0          0       31.13085   -1.82535885 0             
6       20          0       39.17280   -4.76558728 0             
7        0          0       33.36241   -3.76636472 1             
8        0          0       32.62062   -4.22619802 0             
9        0          0       32.71110   -5.14671181 0             
10       0          0       30.62699   -1.25705061 0             
11       0         62       39.20952   -7.03413939 0             
12     200       1062       35.77026  -10.57417468 0             
13       0          0       33.79811   -3.29019380 1             
14       0       1368       37.09257   -3.18178330 0             
15       0          0       34.36407   -3.62933335 1             
16       0       1645       31.44412   -8.27496163 0             
17     500       1703       34.64244   -9.10618458 0             
18       0       1656       34.56927   -9.08551497 0             
19       0       1162       32.92015   -1.94786801 0             
20     500       1763       34.50897   -9.89441250 0             
21     200       2216       34.47343   -9.59498965 0             
22       0          0       35.14967   -0.00000002 0             
23       0       1510       34.58690   -8.98001429 0             
24     500        672       37.94003   -4.12559468 0             
25       0       1645       29.74707   -4.49556758 0             
26       0       1273       37.42275   -3.31753648 0             
27     500        200       39.37078   -9.94253166 0             
28       0          0       31.10426   -1.73536124 0             
29       0       1443       37.61113   -3.26352630 0             
30       0       1256       37.06169   -3.19198926 0             
...   ...        ...        ...       ...          ...           
59371   200      1154       30.05873   -4.90263299 0             
59372     0         0       33.58641   -3.04036855 0             
59373     0         0       32.05855   -5.05210104 0             
59374  5000      1137       32.12502   -6.78435778 0             
59375     0         0       35.14967   -0.00000002 0             
59376 40000        54       39.04694   -5.06149401 0             
59377     0      1581       34.82104   -5.07625759 1             
59378     0         0       33.43701   -4.42359381 1             
59379   500      2101       34.62039   -9.35840452 0             
59380     0         0       31.61328   -1.82022635 0             
59381  6000      1439       34.94210   -9.84919761 0             
59382     0         0       38.83500   -4.88020370 0             
59383     6      1383       37.45476   -3.32359911 0             
59384     0        11       39.62601  -10.05749942 0             
59385     0         0       30.66780   -2.48370953 0             
59386   500      1327       33.95168   -2.02185406 0             
59387     0      1786       31.73879   -8.53201255 0             
59388   100        25       39.17648   -6.95709837 0             
59389     0      1414       30.69240   -3.59382702 0             
59390     0       783       30.64649   -7.36541790 0             
59391     0      1715       31.37085   -8.25816008 0             
59392     0       540       38.04407   -4.27221758 0             
59393     0         0       33.00944   -8.52088818 0             
59394     0         0       33.86685   -4.28740983 1             
59395   500       351       37.63405   -6.12482968 0             
59396    10      1210       37.16981   -3.25384746 0             
59397  4700      1212       35.24999   -9.07062880 0             
59398     0         0       34.01709   -8.75043433 0             
59399     0         0       35.86132   -6.3785

# Model selection

## Model 1

## Model 2

## Model 3

#### Example of simple equation
\begin{equation}
e = mc^2
\end{equation}

#### Example of matrix equation - Cross product formula:

\begin{equation*}
\mathbf{V}_1 \times \mathbf{V}_2 =  \begin{vmatrix}
\mathbf{i} & \mathbf{j} & \mathbf{k} \\
\frac{\partial X}{\partial u} &  \frac{\partial Y}{\partial u} & 0 \\
\frac{\partial X}{\partial v} &  \frac{\partial Y}{\partial v} & 0
\end{vmatrix}
\end{equation*}

#### Example of multiline equation - The Lorenz Equations:

\begin{align}
\dot{x} & = \sigma(y-x) \\
\dot{y} & = \rho x - y - xz \\
\dot{z} & = -\beta z + xy
\end{align}

#### Example of Markdown Table:

| This | is   |
|------|------|
|   a  | table|


In [28]:
N<-nrow(X_final)    #Number of examples
n<-ncol(X_final)    #Number of input variables

names(X_final)<-make.names(names(X_final))

In [32]:

DS<-cbind(X_pca,label=Y)
names(DS)<-make.names(names(DS))
DS$status_group<-factor(DS$status_group)

In [33]:
DS

PC1         PC2         PC3         PC4         PC5         PC6       
1     -5694.65598  592.626326 -186.203084   2.3411072  -1.6078650  3.1386680
2       303.63567  743.109148  -21.352283  13.0292677   3.3197694 -3.2369737
3       292.07535   34.510625   66.124405  12.2689017   2.6605354 -0.2701010
4       325.52024 -413.856576  -53.723258  -5.8008588 -57.5539431  0.8034112
5       330.59903 -682.621009  -67.889513  -0.7626552   6.5200030 -4.5858135
6       310.59983 -682.846965  -66.865744  12.8741376  -0.2025347  1.2369008
7       330.59909 -682.626383  -67.883428  -0.5602538   4.2262372 -2.0422393
8       330.59909 -682.626880  -67.883551  -0.5672506   4.2629810 -2.1878232
9       330.59927 -682.633044  -67.876729  -0.3394667   1.1867450 -1.4623933
10      330.59911 -682.622787  -67.888969  -0.7596641   6.5975057 -5.4860442
11      328.50206 -564.928336  262.356927  17.3122364 -34.4755079 -0.3535044
12      110.06515  402.018810    4.449911  -9.6011791  -1.3814767  4.2817368
13      330.59904 -682.624760  -67.885426  -0.6145265   5.2439510 -2.3354814
14      304.97537  666.725722 -291.466204  12.7442334  -2.3790878 -1.5306315
15      330.59934 -682.633620  -67.876588  -0.2978434   1.2150331 -2.1947632
16      299.24629  972.587427 -140.667651  -5.8800608   0.8141366  0.4266020
17     -201.30131  993.236529 -312.760017 -18.4974719  -2.9332810  2.6106993
18      299.44861  958.805656 -290.443572 -18.5027826  -2.7662498  2.5992042
19      306.11650  627.556984  727.774208   1.2477458   5.3924766 -4.1761136
20     -202.33291 1046.833288 -356.136244  -4.4325995  -3.0812360  3.1182554
21       89.12057 1499.781219 -427.654426  11.4327249  -1.9260533  2.8391588
22      330.59913 -682.623282  -67.887235  -0.6810273   6.3793396 -4.4492815
23      301.36710  864.059642   29.454217 -18.7872639  -1.8723125  2.5097126
24     -182.46878    5.152712   29.151265  14.2150172   3.7189958  0.3453189
25      299.36568  965.369998 -184.056459  12.1448236   1.5771162 -3.6556901
26      306.37735  595.843618 -138.789792 -22.5160313  -1.7503170 -0.9915429
27     -173.76351 -452.194740  156.007477  15.7894882 -15.3957444  4.1429776
28      330.59908 -682.622746  -67.888750  -0.7531586   6.5278068 -4.8488535
29      303.57051  740.704377 -303.798812   3.3037473   2.6241625 -0.6639339
30      307.07318  556.263239 -273.084346   5.7693908  -2.6504025 -1.5116652
...   ...         ...         ...         ...         ...         ...       
59371    109.0192   451.8891  -256.27681   -0.7305062   2.4495077 -2.9662568
59372    330.5991  -682.6251   -67.88548   -0.6241267   5.2791451 -2.7213212
59373    330.5991  -682.6287   -67.88123   -0.5041770   3.2327491 -1.7215145
59374  -4690.0611   359.8952  -157.90031   -8.9235629   2.5652355 -0.4684012
59375    330.5991  -682.6233   -67.88733   -0.6696343   6.3829692 -4.4793092
59376 -39664.9049 -1286.2451   539.68311    7.1421558   3.0578829 -0.3350063
59377    300.9863   876.7994  -326.45985   -6.6538524   0.5355018 -0.7427164
59378    330.5989  -682.6229   -67.88602   -0.6618716   5.1665724 -1.3023644
59379   -208.7589  1385.9391  -377.09125   11.4036230  -1.5813860  2.7109647
59380    330.5992  -682.6250   -67.88547   -0.6033234   4.4907672 -4.5139896
59381  -5695.4132   631.2673  -252.44681    2.3889870  -1.8364712  3.1310981
59382    329.6204  -623.5413   287.23958   -4.4494798  -0.9087705  1.1004918
59383    298.6953   681.4121  -293.93346   11.5210931   0.4360642 -0.9414862
59384    328.2212  -540.5787   719.59340   15.3612251 -52.8801903  0.3960311
59385    330.6013  -682.6846   -67.82727    1.5055300 -22.2556619 -7.0955036
59386   -194.7086   649.5007   -88.24431   14.2731749   1.5003073 -3.9616057
59387    294.4292  1242.9792   625.34110   11.2870491   3.2183212  0.7387443
59388    229.8764  -643.4425    26.69784    3.3570579   4.9812195  3.7412218
59389    302.7567   794.0148   193.18398   -2.2855557   3.6235469 -3.7681889
59390    311.8544   335.8456  1283.22213   10.8808949   5.7895723 -0.5982909
59

In [39]:
data.frame(model$variable.importance)# for rpart


ERROR: Error in data.frame(model$variable.importance): objet 'model' introuvable


In [37]:
#install.packages("rpart.plot")
prp(model)
rpart.plot(model)

ERROR: Error in prp(model): impossible de trouver la fonction "prp"


In [38]:
# plot tree
plot(model, uniform=TRUE,
   main="Classification Tree - Rpart")
text(model, use.n=TRUE, all=TRUE, cex=.8)

ERROR: Error in plot(model, uniform = TRUE, main = "Classification Tree - Rpart"): objet 'model' introuvable


# nnet

In [ ]:
accuracyFromConfusionMatrix <- function(confusion_matrix){
    if(dim(confusion_matrix)[1] < 2){ 
       if(rownames(confusion_matrix) == "nonspam"){
           confusion_matrix <- rbind(confusion_matrix,c(0,0))
           rownames(confusion_matrix)[2] <- "spam"
       }
       if(rownames(confusion_matrix) == "spam"){
           confusion_matrix <- rbind(c(0,0),confusion_matrix)
           rownames(confusion_matrix)[1] <- "nonspam"
       }
    }
    return((confusion_matrix[1,1]+confusion_matrix[2,2])/sum(confusion_matrix))
}


computeResults <- function(Y,Y_hat,explanatory_string="",verbose=TRUE){
    
    confusion_matrix <- table(Y_hat,Y)
    accuracy <- accuracyFromConfusionMatrix(confusion_matrix)
    misclassification_rate <- 1 - accuracy
    
    if(verbose){
        print(paste("[INFO] - Confusion matrix",explanatory_string,":"))
        print(confusion_matrix)
        print(paste("[INFO] - Accuracy",explanatory_string,":",accuracy))
        print(paste("[INFO] - Misclassification rate",explanatory_string,":",misclassification_rate))
    }
    
    return(accuracy)
}

In [ ]:
#test without cross validation

CV_folds <- 10

size_CV <-floor(N/CV_folds)
    
idx_ts<-(((i-1)*size_CV+1):(i*size_CV))  ### idx_ts represents the indices of the test set the i-th fold
X_ts<-X_final[idx_ts,]  
Y_ts<-Y[idx_ts,]
Y_onehot<-labels_onehot[idx_ts,]

idx_tr<-setdiff(1:N,idx_ts) ### idx_tr represents  indices of the training sefor the i-th fold
X_tr<-X_final[idx_tr,]
Y_tr<-Y[idx_tr,]         

DS<-cbind(X_tr,status_group=Y_tr)
DS$status_group<-factor(DS$status_group)
    
    
model<- nnet(status_group~.,data=DS, size=10, maxit=1000, trace=T, MaxNWts=2000)#1000 max iterations ?
Y_hat_ts<- predict(model,X_ts)

In [ ]:
function(x){
   (x-min(x))/(max(x)-min(x))
}

Y_hat <- t(apply(Y_hat_ts, 1, function(x)(x-min(x))/(max(x)-min(x))))
                 
Y_hat[,1] <- ifelse(Y_hat[,1] < 0.99,0,1)
Y_hat[,2] <- ifelse(Y_hat[,2] < 0.99,0,1)
Y_hat[,3] <- ifelse(Y_hat[,3] < 0.99,0,1)
Y_hat
Y_onehot

In [ ]:
Y_onehot<-Y_onehot[,setdiff(colnames(Y_onehot),"id")]
threshold <- 0.5
Y_hat <- ifelse(Y_hat_ts[,1] > threshold,"functional",ifelse(Y_hat_ts[,2] > threshold,"functional_needs_repair","non_functional"))
data.matrix(Y_hat)

In [ ]:
#table(data.frame(Y_hat),data.frame(Y))

require(caret)
#acc <- confusionMatrix(as.factor(Y_hat),as.factor(Y[idx_ts,]))$overall['Accuracy']
confusionMatrix(as.factor(Y_hat),as.factor(Y[idx_ts,]))

In [ ]:
table(Y)

In [ ]:
#change std and error, maybe cross validating and taking the best performing one ?

#nnet model

CV_folds <- 10#10

size_CV <-floor(N/CV_folds)

CV_err<-matrix(0,nrow=1,ncol=CV_folds)


for (i in 1:CV_folds) {
    
    idx_ts<-(((i-1)*size_CV+1):(i*size_CV))  ### idx_ts represents the indices of the test set the i-th fold
    X_ts<-X_pca[idx_ts,]  
    Y_ts<-Y[idx_ts,]
    Y_onehot<-labels_onehot[idx_ts,]
     
    idx_tr<-setdiff(1:N,idx_ts) ### idx_tr represents  indices of the training sefor the i-th fold
    X_tr<-X_pca[idx_tr,]
    Y_tr<-Y[idx_tr,]         
        # Create a dataset including only the first nb_components principal components
    DS<-cbind(X_tr,status_group=Y_tr)
    DS$status_group<-factor(DS$status_group)

    # Model fit (using lm function)
    model<- nnet(status_group~.,data=DS, size=5, decay=1.0e-5, maxit=50, trace=FALSE)#1000 max iterations ?
    #model<- rpart(status_group~.,data=DS)

    # Model predict
    Y_hat_ts<- predict(model,X_ts)
    #Y_hat_ts <- dummy.data.frame(data.frame(Y_hat_ts), sep="_") #for random forest
} 
#print(paste("CV error=",round(apply(CV_err,1,mean),digits=4), " ; std dev=",round(apply(CV_err,1,sd),digits=4)))


#model<- nnet(status_group~.,data=DS, size=3, decay=1.0e-5, maxit=50)
#Y_hat_ts<- predict(model,X_pca)

In [ ]:
Y_hat_ts

# Alternative models





# Conclusions